packages/styling/src/styles/getGlobalClassNames.ts

```diff
@@ -1,17 +1,40 @@
 import { ITheme } from '../interfaces/index';
 
+import { Stylesheet } from '@uifabric/merge-styles';
+import { memoizeFunction } from '@uifabric/utilities';
+
 export type GlobalClassNames<IStyles> = Record<keyof IStyles, string>;
 
 /**
  * Checks for the `disableGlobalClassNames` property on the `theme` to determine if it should return `classNames`
+ * Note that calls to this function are memoized.
  *
+ * @param classNames The collection of global class names that apply when the flag is false. Make sure to pass in
+ * the same instance on each call to benefit from memoization.
  * @param theme The theme to check the flag on
- * @param classNames The global class names that apply when the flag is false
+ * @param disableGlobalClassNames Optional. Explicitly opt in/out of disabling global classnames. Defaults to false.
  */
-export function getGlobalClassNames<T>(classNames: GlobalClassNames<T>, theme: ITheme): Partial<GlobalClassNames<T>> {
-  if (theme.disableGlobalClassNames) {
-    return {};
-  }
+export const getGlobalClassNames: <T>(
+  classNames: GlobalClassNames<T>,
+  theme: ITheme,
+  disableGlobalClassNames?: boolean
+) => Partial<GlobalClassNames<T>> = memoizeFunction(
+  <T>(
+    classNames: GlobalClassNames<T>,
+    theme: ITheme,
+    disableGlobalClassNames?: boolean
+  ): Partial<GlobalClassNames<T>> => {
+    const styleSheet = Stylesheet.getInstance();
 
-  return classNames;
-}
+    if (disableGlobalClassNames || (disableGlobalClassNames === undefined && theme.disableGlobalClassNames)) {
+      // disable global classnames
+      return Object.keys(classNames).reduce((acc: {}, className: string) => {
+        acc[className] = styleSheet.getClassName('__global-class-name__');
```

Comment:
I have a suggestion here; rather than use a very undebuggable `__global-class-name`, why not provide the original value (e.g. `ms-Button`), which will then just append the counter to that. Then you retain the debuggability of the names, while ensuring uniqueness.